### PR TITLE
Fix search of RPC headers

### PR DESCRIPTION
**Pull request by ismaell**
Use RPC_CFLAGS when searching for headers.

Fixes: 94aa9a1b704a ("configure: Allow alternative RPC implementations")
Ref: https://github.com/openSUSE/xinetd/issues/27

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -132,14 +132,17 @@ AS_IF([test x"$with_rpc" != "xno"], [
...
         AC_CHECK_HEADERS([rpc/rpc.h],, [have_rpc=no; break])
         dnl FIXME do a linking test
         AS_IF([test x"$have_rpc" = xmaybe], [
             dnl XXX is this header really optional?
             AC_CHECK_HEADERS([rpc/rpcent.h])
             have_rpc=yes
-            break
         ])
+        CFLAGS="$save_CFLAGS"
+        test x"$have_rpc" = xyes && break
```
Comment (vapier) [inline configure.ac:145]: use `AS_IF` here too please

**Review by vapier — CHANGES_REQUESTED** [review]

File: configure.ac (hunk 1 above)
Comment (ismaell) [reply configure.ac:145]: any reason for using `if` there?

Comment (vapier) [reply configure.ac:145]: `AS_IF` provides more structure to the code, and avoids unintended side-effects with `&&` & `||` usage.  it lends itself to more robust code without having to think as hard about it.

**Comment by davide125** [comment]
config.log: https://gist.github.com/davide125/0da5ccc31c3aa08ff563fa71b5182d13

**Comment by ismaell** [comment]
> prob because this uses `CFLAGS` instead of `CPPFLAGS` that i suggested in your bug report

That's a bug in `autoconf` I'm afraid... what really matters is what the compiler thinks... there must be some workaround.

My guess is that a check was introduced only because in the past `AC_CHECK_HEADERS` (wrongly) used the preprocessor instead of the compiler.

**Comment by vapier** [comment]
> That's a bug in autoconf I'm afraid

it is not a bug in autoconf, it is a bug in your patch.  preprocessor flags go in CPPFLAGS, not CFLAGS.

**Comment by ismaell** [comment]
[And it's documented here](https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Present-But-Cannot-Be-Compiled.html#Present-But-Cannot-Be-Compiled). Anyway, found the actual bug preventing the build from succeeding.

**Comment by ismaell** [comment]
> > That's a bug in autoconf I'm afraid
> 
> it is not a bug in autoconf, it is a bug in your patch. preprocessor flags go in CPPFLAGS, not CFLAGS.

No, the output from `pkg-config` isn't supposed to be added to `CPPFLAGS`, and the autoconf headers check macro should never have used `cpp`.

**Comment by ismaell** [comment]
From the autoconf 2.69 manual: 
> The most important guideline to bear in mind when checking for features is to mimic as much as possible the intended use. Unfortunately, old versions of AC_CHECK_HEADER and AC_CHECK_HEADERS failed to follow this idea, and called the preprocessor, instead of the compiler, to check for headers. As a result, incompatibilities between headers went unnoticed during configuration, and maintainers finally had to deal with this issue elsewhere.
>
>The transition began with Autoconf 2.56. As of Autoconf 2.64 both checks are performed, and configure complains loudly if the compiler and the preprocessor do not agree. However, only the compiler result is considered.

**Comment by ismaell** [comment]
Also, note from `autoconf` 2.70, it doesn't try the preprocessor check at all. Ref: https://www.gnu.org/software/autoconf/manual/autoconf-2.71/html_node/Present-But-Cannot-Be-Compiled.html

**Comment by ismaell** [comment]
> > > That's a bug in autoconf I'm afraid
> > 
> > 
> > it is not a bug in autoconf, it is a bug in your patch. preprocessor flags go in CPPFLAGS, not CFLAGS.
> 
> No, the output from `pkg-config` isn't supposed to be added to `CPPFLAGS`, and the autoconf headers check macro should never have used `cpp`.

`pkg-config` itself doesn't have provisions for `CPPFLAGS`, there's only some distinction for include flags vs anything else, and manually filtering misses the semantic differences anyway, so there's no real fix other than avoiding `CPPFLAGS` when dealing with `pkg-config` output.

**Comment by vapier** [comment]
> > > That's a bug in autoconf I'm afraid
> > 
> > it is not a bug in autoconf, it is a bug in your patch. preprocessor flags go in CPPFLAGS, not CFLAGS.
> 
> No, the output from `pkg-config` isn't supposed to be added to `CPPFLAGS`, and the autoconf headers check macro should never have used `cpp`.

that is incorrect.  like i said, preprocessor flags go in CPPFLAGS, not CFLAGS.  i think you're confused by the two -- when invoking the compiler, it uses both CPPFLAGS & CFLAGS, not just CFLAGS.

**Comment by vapier** [comment]
> `pkg-config` itself doesn't have provisions for `CPPFLAGS`, there's only some distinction for include flags vs anything else, and manually filtering misses the semantic differences anyway, so there's no real fix other than avoiding `CPPFLAGS` when dealing with `pkg-config` output.

that's because `pkg-config` only includes flags that are suitable for all frontends.  `CFLAGS` is specific to C, it doesn't work with C++ nor the preprocessor.  `CPPFLAGS` is used by all of them.

**Comment by iluceno** [comment]
Ping.

**Comment by vapier** [comment]
the answer is still the same.  the code should be using CPPFLAGS, not CFLAGS.
https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Preset-Output-Variables.html

i just went ahead and fixed it myself since there were other bugs in the rpc rewrite.

**Comment by ismaell** [comment]
I'm unhappy about commit 4ada39f9eb615af0cb9b179a36bcadeb96de2357, after months of no feedback it replaces this work and it omits any attribution...

Also it didn't fix the calls to `AC_CHECK_HEADERS` for autoconf 2.69.

**Comment by ismaell** [comment]
> the answer is still the same. the code should be using CPPFLAGS, not CFLAGS. 

I agree that it's improper to use `CFLAGS`, but are you going to fix the `pkgconf`/`pkg-config` side of the problem?

**Comment by vapier** [comment]
you had months to address the feedback. we needed a fix, so if you weren't willing to do it, someone had to as the tree was broken.

if there's a problem with the current git, feel free to send a PR. i don't know what AC_CHECK_HEADERS issue you mean.

**Comment by ismaell** [comment]
> you had months to address the feedback. we needed a fix, so if you weren't willing to do it, someone had to as the tree was broken.

The use of `CPPFLAGS` works in this case, but is improper for `pkg-config`. That went unanswered since mid March.
 
> if there's a problem with the current git, feel free to send a PR. i don't know what AC_CHECK_HEADERS issue you mean.

The problem is the lack of attribution for something that is almost identical.

**Comment by ismaell** [comment]
The problem with `AC_CHECK_HEADERS` is that the 4th argument required for `autoconf` 2.69. I'll make a new PR for that.

**Comment by vapier** [comment]
> > you had months to address the feedback. we needed a fix, so if you weren't willing to do it, someone had to as the tree was broken.
> 
> The use of `CPPFLAGS` works in this case, but is improper for `pkg-config`. That went unanswered since mid March.

i don't know what you mean by improper.  if you want to argue with the pkg-config projects to rename their flags, go for it.  but how they choose to name their options is irrelevant.  `CPPFLAGS` is the correct variable name for integrating with build systems, not `CFLAGS`.

> > if there's a problem with the current git, feel free to send a PR. i don't know what AC_CHECK_HEADERS issue you mean.
> 
> The problem is the lack of attribution for something that is almost identical.

if you mean i took your patch & credit for your work, then not really.  when you refused to write the code correctly, i did it from scratch based on the bug.  i'll note that i triaged & diagnosed the bug, and suggested the fix -- you took my suggestion and wrote a patch, albeit didn't note the origin of the idea in your commits.  not that i care (i didn't complain during the review), but you're insisting on this level of nitpicking, so i feel obliged to highlight.

that the changes look similar is a matter of the fixes being straightforward.  and that i suggested the fix in the first place that we both implemented.  the Makefile.am changes were trivial to detect by attempting to compile the package against libtirpc.

further, if you look at your PR, it's buggy wrt `break` calls.  my change doesn't have that since i didn't base it on yours.

> The problem with `AC_CHECK_HEADERS` is that the 4th argument required for `autoconf` 2.69. I'll make a new PR for that.

f i really had taken your work, i probably would have blindly copied this logic.  the code i committed worked in my testing, and i don't know what you're referring to here, so feel free to send a PR explaining it.

**Comment by vapier** [comment]
> I agree about that, but it isn't the way it is used, and it isn't the way pkg-config defines it. It explicitly defines it in the documentation as pre-processor and compiler flags.

any .pc file including compiler flags (i.e. non-preprocessor flags) is broken.  the only valid use of pkg-config flags in the eocsystem, both for export & for consumption, is preprocessor flags i.e. CPPFLAGS.

> The break calls are proper usage to short-circuit the header search.

that would be true iff multiple headers were passed to `AC_CHECK_HEADERS`.  there aren't.

> It's required by some compatibility macros in autoconf 2.69 and earlier, if you're using a newer version you wouldn't see the error.

that's incorrect.  you're assuming setting of CFLAGS, not CPPFLAGS.  since i set the right variable, there are no warnings.

**Comment by ismaell** [comment]
<...>
> any .pc file including compiler flags (i.e. non-preprocessor flags) is broken. the only valid use of pkg-config flags in the eocsystem, both for export & for consumption, is preprocessor flags i.e. CPPFLAGS.

Sensible as it might be, since real-world usage and documentation contradict that, it is unsafe to assume such position.

<...>
> > It's required by some compatibility macros in autoconf 2.69 and earlier, if you're using a newer version you wouldn't see the error.
> 
> that's incorrect. you're assuming setting of CFLAGS, not CPPFLAGS. since i set the right variable, there are no warnings.

The point is to have the same behavior on 2.69 as on later versions, for consistency. Also, while preprocessor errors will affect the compiler, the other way around doesn't work (and without this the test succeeds in 2.69 while fails on newer versions).

**Comment by vapier** [comment]
> <...>
> > any .pc file including compiler flags (i.e. non-preprocessor flags) is broken. the only valid use of pkg-config flags in the eocsystem, both for export & for consumption, is preprocessor flags i.e. CPPFLAGS.
> 
> Sensible as it might sound, since real-world usage and documentation contradict that, it is unsafe to assume such position.

it's unrealistic to assume otherwise. unless you're arguing that only C is supported, and C++ cannot be used.

> <...>
> > > It's required by some compatibility macros in autoconf 2.69 and earlier, if you're using a newer version you wouldn't see the error.
> > 
> > that's incorrect. you're assuming setting of CFLAGS, not CPPFLAGS. since i set the right variable, there are no warnings.
> 
> The point is to have the same behavior on 2.69 as on later versions, for consistency. Also, while preprocessor errors will affect the compiler, the other way around doesn't work (and without this the test succeeds in 2.69 while fails on newer versions).

the code, as written in the tree, works with every version of autoconf because i used the right variable (CPPFLAGS). it will continue to work correctly with future versions of autoconf. disabling preprocessor checks with <=autoconf-2.69 is basically an optimization. I'm not against it, but it's certainly not necessary or fixing any known bugs.

**Comment by ismaell** [comment]
<...>
> it's unrealistic to assume otherwise. unless you're arguing that only C is supported, and C++ cannot be used.

Probably should be added to both...

<...>
> > The point is to have the same behavior on 2.69 as on later versions, for consistency. Also, while preprocessor errors will affect the compiler, the other way around doesn't work (and without this the test succeeds in 2.69 while fails on newer versions).
> 
> the code, as written in the tree, works with every version of autoconf because i used the right variable (CPPFLAGS). it will continue to work correctly with future versions of autoconf. disabling preprocessor checks with <=autoconf-2.69 is basically an optimization. I'm not against it, but it's certainly not necessary or fixing any known bugs.

It's possible for a header to be usable from the ponit of view of preprocessor yet not usable by the compiler, which makes the two behaviors not at all equivalent.

**Comment by vapier** [comment]
> It's possible for a header to be usable from the point of view of preprocessor yet not usable by the compiler, which makes the two behaviors not at all equivalent.

yes, that is the theory.  in practice, i doubt it actually comes up.  further, as i stated, it isn't fixing any known bugs, nor is it changing any actual behavior since autoconf-2.69 ignores (other than a warning) the preproc result.

**Pull request closed without merging**